package/cpp/UTFConversion.h

```diff
@@ -0,0 +1,147 @@
+#include <cstddef>
+#include <cstdint>
+#include <string>
+#if defined(__ARM_NEON) || defined(__ARM_NEON__)
+#include <arm_neon.h>
+#endif
+
+// A super-optimized UTF16-to-UTF8 converter for ARM/mobile.
+// It converts UTF16 input (given as a const char16_t* and length)
+// into a std::string in UTF8. Two NEON fast-paths are used:
+//
+//   • If a block of 8 UTF16 code units are all in the range [0x0800,0xFFFF]
+//     (and not surrogates), each produces exactly 3 UTF8 bytes.
+//     We compute the three bytes in parallel and use vst3_u8 to store 24 bytes.
+//
+//   • If a block of 8 code units are all in the range [0x80,0x07FF] (i.e. none
+//     are ASCII, so all are 2-byte UTF8), we use a similar approach with vst2_u8.
+//
+// In any other case (including surrogate pairs or mixed ranges) we fall back
+// to a scalar loop that handles surrogate pairs and variable-length conversion.
+std::string utf16_to_utf8(const char16_t* input, size_t len) {
+  // Preallocate worst-case output: every code unit might be 3 bytes.
+  std::string out;
+  out.resize(len * 3);
+  char* dest = &out[0];
+  const char16_t* src = input;
+  const char16_t* end = input + len;
+
+#if defined(__ARM_NEON) || defined(__ARM_NEON__)
+  // Preload frequently used constants into NEON registers.
+  const uint16x8_t v_d800 = vdupq_n_u16(0xD800);
+  const uint16x8_t v_dfff = vdupq_n_u16(0xDFFF);
+  const uint16x8_t v_thr = vdupq_n_u16(0x0800);
+  const uint16x8_t v_mask3F = vdupq_n_u16(0x3F);
+  const uint16x8_t v_constE0 = vdupq_n_u16(0xE0); // for 3-byte header: 0xE0
+  const uint16x8_t v_constC0 = vdupq_n_u16(0xC0); // for 2-byte header: 0xC0
+  const uint16x8_t v_const80 = vdupq_n_u16(0x80); // for continuation bytes
+
+  while (src + 8 <= end) {
+    // Load 8 UTF16 code units.
+    uint16x8_t vec = vld1q_u16(reinterpret_cast<const uint16_t*>(src));
+
+    // Check for any surrogates (i.e. values in [0xD800, 0xDFFF]).
+    uint16x8_t ge_d800 = vcgeq_u16(vec, v_d800);
+    uint16x8_t le_dfff = vcleq_u16(vec, v_dfff);
+    uint16x8_t surrogate = vandq_u16(ge_d800, le_dfff);
+    if (vmaxvq_u16(surrogate) != 0) {
```

Comment:
Is this available on armv7 NEON?

Reply:
It indeed isn't. Need to implement manually